### ISSUE TITLE
Error with Google Photo

### ISSUE THREAD

**Issue by Ludovicis** [nextcloud/integration_google#5]
{"reqId":"d3sVyJhXtBtjunMr9PGT","level":2,"time":"2020-10-02T15:13:16+00:00","remoteAddr":"x","user":"x","app":"integration_google","method":"GET","url":"/apps/integration_google/photo-number","message":"Google API error : Client error: `GET https://photoslibrary.googleapis.com/v1/mediaItems?pageSize=100` resulted in a `403 Forbidden` response:\n{\n  \"error\": {\n    \"code\": 403,\n    \"message\": \"Photos Library API has not been used in project 152656787343 before or i (truncated...)\n","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}

**Comment by eneiluj** [comment]
@Ludovicis Could you try to "disconnect from Google" in Nextcloud settings, then logout from your Google account, and then try to reconnect in Nextcloud settings. This will ensure you get a new OAuth token.

I suspect you first connected to your Google account in Nextcloud while the photo API was not yet added in your Google API application...if that makes any sense...:smile:

**Comment by Ludovicis** [comment]
With 0.0.2 version google-integration
Ok i have add api, i must add scope also  on my app on OAuth consent screen, please? 
With this version i can no more connect on my google account, the google connect account from /settings/user/migration redirect to user/connected-accounts after google account connection

{"reqId":"iQpmacGHVRNkNdOnlxge","level":2,"time":"2020-10-02T19:15:51+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"integration_google","method":"GET","url":"/apps/integration_google/photo-number","message":"Google API error : Client error: `GET https://photoslibrary.googleapis.com/v1/mediaItems?pageSize=100` resulted in a `403 Forbidden` response:\n{\n  \"error\": {\n    \"code\": 403,\n    \"message\": \"Photos Library API has not been used in project 152656787343 before or i (truncated...)\n","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}
{"reqId":"5lPfeXGgYJ5ciq6gRk8w","level":3,"time":"2020-10-02T19:16:13+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=q8vihferkd&code=4%2F4wGeHJWoJfb0pTJh2yuCOUrD2_fC0qrnhjvQCo0hCLbGbNXmynxtorr5I2LgGpTp2bOyGyOnc6L90h9kmKLmXBw&scope=profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}
{"reqId":"c5OjbhiIc2hacaAM8ihc","level":3,"time":"2020-10-02T19:16:25+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=fcwb18sfks&code=4%2F4wF5pljNleBjdjg6IuMTJyXNzR58_e7B7D7KS7EGMvcKrMkB0cJCQ5mnBEi1jpoZbFq7JSpxaAf4UmFkGk0_Ny0&scope=profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly+openid&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}
{"reqId":"xe2qZtx50MSqieUZapmb","level":3,"time":"2020-10-02T19:16:37+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=v70oj3j7r&code=4%2F4wGUgl8wsWQ2Bw0ghUDS8LXr69tt-bPjYby6_ZqnXmeMXqPYkXQ6jAPmYA-y4LwKmVk49cNT36jgB1PvDq6Fo0A&scope=profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}
{"reqId":"mkzKaETgVfVFQtDxwqSB","level":3,"time":"2020-10-02T19:17:21+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=1azmxk9gsc&code=4%2F4wGVxGZEgPZ3ityXANU9Y-QRORYjoSdKNKaFr3yFtJM1pL8Ssb2jQphpo7gm7_8Q57EaiBA-q8QFFY2NxGQhcOI&scope=profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36","version":"20.0.0.8"}
{"reqId":"RjiVxcocZUSev67XHUKD","level":3,"time":"2020-10-02T19:18:31+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=bfqq06hy28&code=4%2F4wGL2cjgw20xfG2l4YwBywtYw_JTb90QVKVEWESFEblTNgWSxlfnmWHBHkXA-LRiyWRGXgFNNoAomPGPTrTi_dE&scope=profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly+openid&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:81.0) Gecko/20100101 Firefox/81.0","version":"20.0.0.8"}
{"reqId":"nrquZVHtRMGJNFzYXedD","level":3,"time":"2020-10-02T19:18:45+00:00","remoteAddr":"X.X.X.X","user":"X.X","app":"PHP","method":"GET","url":"/apps/integration_google/oauth-redirect?state=lcg8f5ooxa&code=4%2F4wG_YjqFUDZZ5LLPyQMrKh6uSvPifVuty89tJmlPz5pYwr4DAAPoJvTF8XUBecinBiP-k-c31Vl3OcPaRZIzm1A&scope=profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fphotoslibrary.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.readonly+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcalendar.events.readonly+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcontacts.readonly&authuser=0&prompt=none","message":"Undefined index: error at /var/www/html/nextcloud/apps/integration_google/lib/Controller/ConfigController.php#169","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:81.0) Gecko/20100101 Firefox/81.0","version":"20.0.0.8"}

**Comment by eneiluj** [comment]
The errors you get are not the problem's source. You get that because there is a bug (that will be fixed) but also it can only happen when Google's answer to authentication is wrong which means it only gives you a partial answer (refresh token is missing).

It's not about the app version. That happened to me too. To make sure it works, you need to logout of your google account (on Google's website) to make sure you get a fresh new token next time you connect to Google in Nextcloud connected accounts settings.

I'll check if I find a way to avoid this issue by telling Google that we need the refresh token each time we authenticate.

**Comment by jakobroehrl** [comment]
I have the same problem.
When I open the settings page I get this errors:

Fehler beim Abrufen der Anzahl der Google-Fotos: "Client error: `GET https:\/\/photoslibrary.googleapis.com\/v1\/mediaItems?pageSize=100` resulted in a `403 Forbidden` response:\n{\n \"error\": {\n \"code\": 403,\n \"message\": \"Photos Library API has not been used in project 954837144118 before or i (truncated...)\n"✖

Fehler beim Abrufen der Anzahl der Google-Kontakte: "Client error: `GET https:\/\/people.googleapis.com\/v1\/people\/me\/connections?personFields=names&pageSize=100` resulted in a `403 Forbidden` response:\n{\n \"error\": {\n \"code\": 403,\n \"message\": \"People API has not been used in project 954837144118 before or it is dis (truncated...)\n"

**Comment by eneiluj** [comment]
@jakobroehrl Are you sure you added the APIs in your Google OAuth app (library)?

**Comment by eneiluj** [comment]
If you add them, then you need to disconnect and reconnect to Google in Nextcloud migration settings.

**Comment by jakobroehrl** [comment]
> @jakobroehrl Are you sure you added the APIs in your Google OAuth app (library)?
Yes I added it and I can import contacts and calendars

> If you add them, then you need to disconnect and reconnect to Google in Nextcloud migration settings.
Reconnect doesn't change something

**Comment by eneiluj** [comment]
@jakobroehrl My interpretation of Google's response (in your logs) means that the APIs were added in Google dev console but the token you're using was generated before adding the APIs. To make sure you really get a new authentication token, you have to logout from your Google account (in any Google website) and reconnect in Nextcloud settings. I remember having solved the same problem by doing this.

@Ludovicis Any clue about what action solved the problem on your side?

**Comment by Ludovicis** [comment]
@Ludovicis Any clue about what action solved the problem on your side?
In my memory, I added as you indicated in the wiki the google api (Maybe more api...). I waited a bit for this to be taken into account on the google side, i have also added my domain in Authorized domains on OAuth consent screen on my google app. On Scopes for Google APIs i have email | profile | openid on OAuth consent screen on my google app.
Then I disconnected and reconnected on nextcloud. Everything works like a charm now
